src/librustc_typeck/check/mod.rs

```diff
@@ -5052,7 +5053,31 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             self.tcx.generics_of(*def_id).has_self
         }).unwrap_or(false);
 
-        let def_id = def.def_id();
+        let mut new_def = def;
+        let (def_id, ty) = match def {
```

Comment:
I think this would be clearer as `if let` rather than `match`, because it's really only `SelfCtor` that's being special-cased.

Reply:
Fixed.